gnucash/register/register-gnome/completioncell-gnome.c

```diff
@@ -0,0 +1,917 @@
+            od->count = 1;
+            od->occurrence = box->occurrence;
+            g_hash_table_insert (box->item_hash, g_strdup (menustr), od);
+        }
+        box->occurrence++;
+    }
+}
+
+void
+gnc_completion_cell_set_value (CompletionCell* cell, const char* str)
+{
+    if (!cell || !str)
+
+    gnc_basic_cell_set_value (&cell->cell, str);
+}
+
+static inline void
+list_store_append (GtkListStore *store, char* string,
+                   char* markup, gint weight)
+{
+    GtkTreeIter iter;
+
+    g_return_if_fail (store);
+    g_return_if_fail (string);
+    g_return_if_fail (markup);
+
+    gtk_list_store_append (store, &iter);
+
+    gtk_list_store_set (store, &iter, TEXT_COL, string,
+                                      TEXT_MARKUP_COL, markup,
+                                      WEIGHT_COL, weight, -1);
+}
+
+static char*
+normalize_and_fold (char* utf8_string)
+{
+    char *normalized = g_utf8_normalize (utf8_string, -1, G_NORMALIZE_NFC);
+    if (!normalized)
+        return NULL;
+
+    char *folded = g_utf8_casefold (normalized, -1);
+    g_free (normalized);
+    return folded;
+}
+
+static void
+add_item (gpointer key, gpointer value, gpointer user_data)
+{
+    PopBox* box = user_data;
+    gchar *hash_entry = g_strdup (key);
+
+    if (hash_entry && *hash_entry)
+    {
+        gnc_utf8_strip_invalid_and_controls (hash_entry);
+        gchar *key_norm_fold = normalize_and_fold (hash_entry);
+        gchar *found = g_strstr_len (key_norm_fold, -1, box->newval);
+
+        if (found)
+        {
+            gchar *markup, *prefix = NULL, *match = NULL, *suffix = NULL;
+            gint prefix_len, suffix_len, offset;
+            gint occurrence_difference;
+            gboolean have_boundary = FALSE;
+            glong len = g_utf8_strlen (box->newval, -1);
+            gulong loc = g_utf8_pointer_to_offset (key_norm_fold, found);
+            OccurrenceData *od = (OccurrenceData*)value;
+
+            if (box->register_is_reversed)
+                occurrence_difference = od->occurrence + 1;
+            else
+                occurrence_difference = box->occurrence - od->occurrence;
+
+            if (loc > 0)
+                prefix = g_utf8_substring (hash_entry, 0, loc);
+            else
+                prefix = g_strdup ("");
+            prefix_len = g_utf8_strlen (prefix, -1);
+
+            match = g_utf8_substring (hash_entry, loc, loc + len);
+
+            if ((loc > 1) && g_strcmp0 (g_utf8_substring (hash_entry, loc - 1, loc), " ") == 0)
+               have_boundary = TRUE;
+
+            suffix = g_utf8_substring (hash_entry, loc + len, g_utf8_strlen (hash_entry, -1));
+            suffix_len = g_utf8_strlen (suffix, -1);
+
+            markup = g_markup_printf_escaped ("%s<b>%s</b>%s", prefix, match, suffix);
+
+            if (prefix_len == 0)
+                offset = 0;
+            else
+            {
+                if (have_boundary)
+                    offset = (prefix_len - 1) * -1;
```

Comment:
I originally had 50 for the boundary offset but can not remember why I changed it. It may of been when I combined it with the time factor.

So just the boundary value needs to be bigger than the description text length and in words bigger so 0, 100, 1000, that makes sense.

But when combined with time, this is where I am getting confused.
If you have...
17/05/2021   bank
08/05/2023   Bank of America
18/05/2023   best long text
19/05/2023   the best

and you type 'b', which should be selected and what should the order be?

Reply:
> and you type 'b', which should be selected and what should the order be?

If I might jump in a bit. I think inherently the relationship between time and place (in the string) is always going to be a bit subjective, so it needs to be done one way or another but I don't think there is a fully objective way to look at it, hence whatever weight you give time or place is fine as long as it doesn't completely overwhelm the other side to the point of not even kicking in. You should see things jump back and forth as you adjust one or the other. Let me explain as per your example.

In you example, the first 3 transactions are so close to each other in time that it should almost be irrelevant for them. Then you have bank alone 2 years ago. So intuitively I would say (in order top to bottom) "b" would give me:
18/05/2023 best long text
08/05/2023 Bank of America
19/05/2023 the best
17/05/2021 bank

But if you push "08/05/2023 Bank of America" down in time eventually I would expect to see it jump one line down below "the best". Somewhere between March and the begining of April or so? When is really subjective, there isn't I think an objective way to measure the time-place relationship, it will probably come down to feel. Similarly if "17/05/2021 bank" where to come back up in time, I would expect it to jump above "19/05/2023 the best" even before it gets to that date, but when? Maybe when it arrives around the beginning of 2023 or so? I can't tell you exactly. But I don't think anybody can, or said otherwise everybody is going to have a different opinion. In practice it will depend on how many transactions are entered in a book, but let's not go there. So we need an executive decision on this one I think.

I also think this is the point in the discussion where @jralls is going to start double-rolling his eyes even harder ;-)

Reply:
> I also think this is the point in the discussion where @jralls is going to start double-rolling his eyes even harder ;-)

Too late. That started some time ago. Worse, they keep crossing. ;-)

IMO the number of non-matching characters is irrelevant, arbitrary, and impossible to justify or even explain to users when it's more than 0. So is whether the non-matching characters are before or after the match. I think it would be reasonable to require matches to begin on word boundaries, so in the example set above "ank" or "mer" wouldn't match at all. That's also a bit arbitrary but it can at least be justified as reducing the quantity of matches to a manageable number.

@gjanssens wrote
> The remaining question is how this should then influence the total weight. Should it truly be secondary as in v4 ? 

Time isn't secondary in v4. The match-tree-stuffing algorithm creates a single match string for each letter as it iterates over the account's native-sorted split list, *discarding previous matches*. Since posted date is the primary key on that native sort because it's the account's split list, newer strings replace older ones in the hash table, resulting in the string with the latest posted date that starts with the same input is presented to the user. You type a 'b' and you get the last posted transaction whose description starts with '[Bb]'.

I think we should mostly maintain that behavior: The exact match should be presented first followed by the first n partial matches in reverse date order after the exact match if any where n is some arbitrary number (~15?) to keep the presentation sane. I suppose that in cases of more than one match at the same time (and see below about that) sort based on matches-start-of-string and matches-start-of-some-other-word, and while I don't think it's all that important there are no doubt some obsessives out there who will be upset if there isn't some predictable pattern to same-time matches.

Reply:
> > I also think this is the point in the discussion where @jralls is going to start double-rolling his eyes even harder ;-)
> 
> Too late. That started some time ago. Worse, they keep crossing. ;-)
> 
> IMO the number of non-matching characters is irrelevant, arbitrary, and impossible to justify or even explain to users when it's more than 0.

And yet it's what v4 depended upon, see below.

> So is whether the non-matching characters are before or after the match.

That was initially used as an approximation to prefer start of string matches, but I agree in the current weighing algorithm it doesn't matter. And considering the match part is the same for all proper candidate strings, we could also work with total string size. My only objection is that string size is less expressive than "number of non-matching characters" when describing the quality of a match, again assuming this is what v4 did.

 I think it would be reasonable to require matches to begin on word boundaries, so in the example set above "ank" or "mer" wouldn't match at all. That's also a bit arbitrary but it can at least be justified as reducing the quantity of matches to a manageable number.

I agree that skipping matches that aren't on word boundary is probably a fair compromise even though it goes against the search habits I have picked up while using another accounting program.

> 
> @gjanssens wrote
> 
> > The remaining question is how this should then influence the total weight. Should it truly be secondary as in v4 ?
> 
> Time isn't secondary in v4. The match-tree-stuffing algorithm creates a single match string for each letter as it iterates over the account's native-sorted split list, _discarding previous matches_. Since posted date is the primary key on that native sort because it's the account's split list, newer strings replace older ones in the hash table, resulting in the string with the latest posted date that starts with the same input is presented to the user. You type a 'b' and you get the last posted transaction whose description starts with '[Bb]'.
> 

Again, that's not how I read the code. It's a recursive hash table in which shorter strings take precedence over longer strings in case of same match length. So if you type a b, you'll get the shortest string that starts with '[bB]'. Only if you have two equal-length strings that start with '[bB]', the last posted one will be preferred. So in V4, time is secondary to string size, or considering the match part is the same in all candidate strings, time is secondary to non-matched part.

You can easily verify this with the test transactions suggested above by @Bob-IT : if these four transactions are in your set and you start a new transaction by typing a "B", it would suggest "Bank" in v4. It's not your exact match and even furthest away in time. It's chosen because it's the shortest string that has the proper match prefix.

> I think we should mostly maintain that behavior: The exact match should be presented first followed by the first n partial matches in reverse date order after the exact match if any where n is some arbitrary number (~15?) to keep the presentation sane.

If you want to maintain the v4  behaviour, you have to prefer shorter strings that have a match over longer ones, which is why I added number of mismatching characters in the weight formula. That same formula would also present a true exact match before anything else.

> I suppose that in cases of more than one match at the same time (and see below about that) sort based on matches-start-of-string and matches-start-of-some-other-word, and while I don't think it's all that important there are no doubt some obsessives out there who will be upset if there isn't some predictable pattern to same-time matches.

In the end I have no strong opinion on which sort order is best. The point is we need to get to something that can be sorted on by a GtkTreeModel. You can't just tell that model "if it's an exact match, put this first, otherwise just sort on date posted". The sortable should be something measurable. So we have to put **some** number of value on each of the possible match candidates. I started from how I observed the v4 behaviour and tried to model something that would at least make the first match candidate as close as possible to what it used to be in v4. That is shortness of mismatch is the primary qualifier and only in case of equal mismatch sizes the nearest in time is chosen.

Frankly, I think this is very easily accomplished by creating two columns in the GtkTreeModel: one counting the number of non-matching characters and the second with the post date. Then add two sort models on top of that: the first one sorting on the post date, the second one sorting on the mismatch count column. I do assume here that GtkTreeModel uses a stable sort, which should be verified.

The added complexity comes from the **new** desire to also match on word boundaries, not only on start of string. The question is how important are these matches compared to start of string matches ? Should they **always** come after start of string matches or are some word boundary matches better than start of string matches (again considering v4 clearly preferred shorter strings over longer ones) ? That will make a difference in how to add it into the sorting mix.

Reply:
> You can easily verify this with the test transactions suggested above by @Bob-IT : if these four transactions are in your set and you start a new transaction by typing a "B", it would suggest "Bank" in v4. It's not your exact match and even furthest away in time. It's chosen because it's the shortest string that has the proper match prefix.

Nope:
<img width="1138" alt="Screenshot 2023-05-20 at 09 35 10" src="https://github.com/Gnucash/gnucash/assets/58910/02e97cbf-37f3-4719-975e-11853a4fca46">

Reply:
> Again, that's not how I read the code

Then [read it again](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/libgnucash/app-utils/QuickFill.c#L248). It's called from [add_quickfill_completions](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/gnucash/register/ledger-core/split-register-load.c#L243) in split-register-load.c via [gnc_quickfill_cell_add_completion](https://github.com/Gnucash/gnucash/blob/bba49a6d09ddc004daccadaa14361187f8c66e37/gnucash/register/register-core/quickfillcell.c#LL337C1-L337C34) in quickfillcell.c. Sort is set by quickfillcell's `sort` member that's always [`QUICKFILL_LIFO`](https://github.com/Gnucash/gnucash/blob/bba49a6d09ddc004daccadaa14361187f8c66e37/gnucash/register/register-core/quickfillcell.c#LL280C18-L280C32) as [gnc_quickfill_cell_set_sort](https://github.com/Gnucash/gnucash/blob/bba49a6d09ddc004daccadaa14361187f8c66e37/gnucash/register/register-core/quickfillcell.c#LL314C1-L314C28) is never called.

[`struct _Quickfill`](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/libgnucash/app-utils/QuickFill.c#L35) contains a match string, a length, and a hashtable of children. At each level the hashtable is keyed with a single `gunichar`. At each level the match string is [replaced with the current string's text](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/libgnucash/app-utils/QuickFill.c#L298) before [getting the next codepoint in the string and recursing into the child hashtable](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/libgnucash/app-utils/QuickFill.c#L303).

`add_quickfill_completions` is [called](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/gnucash/register/ledger-core/split-register-load.c#L670) from [`gnc_split_register_load `](https://github.com/Gnucash/gnucash/blob/ba41b1bbf15eaef1d2fadee45b356b65a643cd38/gnucash/register/ledger-core/split-register-load.c#L353) while traversing the register's account's split list.

Reply:
I think you might have gone astray by populating an empty register and then testing without having closed and reopened the register. [`gnc_quickfill_cell_leave`](https://github.com/Gnucash/gnucash/blob/bba49a6d09ddc004daccadaa14361187f8c66e37/gnucash/register/register-core/quickfillcell.c#L246) calls `gnc_quickfill_insert` with the new text, so if you start with an empty register and create some transactions the `Quickfill` will be populated in the order of creating the transactions, ignoring `posted_date`.

Reply:
> You can't just tell that model "if it's an exact match, put this first, otherwise just sort on date posted". 

Sure you can. If it's exact match you tell the model that the date posted is the max of the other date_posted + 1.

Reply:
Aha, I stand corrected. In order to get that result as well, I had to close and reopen the register in question after entering the base transactions first. I neglected to do so twice, so the most recently entered entries got precedence, regardless of their post date in the interim. Apologies for the looking in the wrong direction. (My code reading was clearly superficial as well...).

So with that I agree string length (or mismatch length) doesn't seem to be relevant at all. Only first match closest in time. That should simplify the sorting quite a bit. Remains to answer is how to weigh the on word boundary matches against start of string matches. Start of string is the only thing we had in v4, so there's no answer from that version, other than the most vocal users prefer to keep start of string as most relevant match.

Reply:
> I think you might have gone astray by populating an empty register and then testing without having closed and reopened the register. [`gnc_quickfill_cell_leave`](https://github.com/Gnucash/gnucash/blob/bba49a6d09ddc004daccadaa14361187f8c66e37/gnucash/register/register-core/quickfillcell.c#L246) calls `gnc_quickfill_insert` with the new text, so if you start with an empty register and create some transactions the `Quickfill` will be populated in the order of creating the transactions, ignoring `posted_date`.

Exactly that as you can read in my reply that crossed yours.

Reply:
> > You can't just tell that model "if it's an exact match, put this first, otherwise just sort on date posted".
> 
> Sure you can. If it's exact match you tell the model that the date posted is the max of the other date_posted + 1.

You make my point. That's not telling put the exact match first. That's finding an algorithm that would put that match first. I really wanted to express the GtkTreeModel doesn't accept if/then structures to sort, you have to feed it a column (or multiple) it can sort on, numerically or alphanumerically. With the string length out of the way, your algorithm is fine indeed.

Reply:
> Remains to answer is how to weigh the on word boundary matches against start of string matches.

We could have a preference. 😉

I suspect in normal use it won't matter that much and that getting rid of the matches in the middle of a word will mollify most of the objectors. The loudest and most persuasive complaint is the freezes, which is why I asked @Bob-IT to separate that fix and get it pushed ASAP.

For the cases where there's a start-of-string and a middle-of-string match on the same day I suppose we could subtract  a second from the middle-of-string timestamp.

But there's more "remains to answer": Do we want to re-create the behavior where transactions added in the current session, even if their posted dates are more in the past, sort higher? How do we handle duplicate strings?

For the first the closest replication of v4 would record the register open time and compare that to the transaction's `entered_date`; if `open time < entered_date` then use `entered_date` else use `posted_date`. Alternatively we could always use `entered_date` figuring that most of the time it won't change the order significantly and that nobody will notice when it doesn't.

 For the second I think we should show only the most recent match on a particular string. Absent any differentiators presenting multiple strings (each associated with possibly different transfer accounts or amounts) will only frustrate users.

Reply:
> > Remains to answer is how to weigh the on word boundary matches against start of string matches.
> 
> We could have a preference. wink

Now **I** start rolling my eyes :wink: 

> 
> I suspect in normal use it won't matter that much and that getting rid of the matches in the middle of a word will mollify most of the objectors. The loudest and most persuasive complaint is the freezes, which is why I asked @Bob-IT to separate that fix and get it pushed ASAP.

Understood for the freezes.

> 
> For the cases where there's a start-of-string and a middle-of-string match on the same day I suppose we could subtract a second from the middle-of-string timestamp.

That would be probably be sufficient, so agreed.

> 
> But there's more "remains to answer": Do we want to re-create the behavior where transactions added in the current session, even if their posted dates are more in the past, sort higher? How do we handle duplicate strings?
> 
> For the first the closest replication of v4 would record the register open time and compare that to the transaction's `entered_date`; if `open time < entered_date` then use `entered_date` else use `posted_date`. Alternatively we could always use `entered_date` figuring that most of the time it won't change the order significantly and that nobody will notice when it doesn't.

I'd start with the simplest option (always use `entered_date`) and continue to tweak if there are too many complaints of undesired behaviour due to this choice.

> 
> For the second I think we should show only the most recent match on a particular string. Absent any differentiators presenting multiple strings (each associated with possibly different transfer accounts or amounts) will only frustrate users.

Absolutely agreed. So we need to filter out duplicates from the list, and I'd say case-insensitively.